### DISCUSSION TITLE
AddPickupAndDelivery implicitly constraint node order?

### DISCUSSION THREAD

**Discussion by vuongdanghuy** [google/or-tools#3973] [Routing (and legacy CP) questions]
**What version of OR-Tools and what language are you using?**
Version: v9.7
Language: Java/Python

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
Routing solver

**What operating system (Linux, Windows, ...) and version?**
Linux Mint 21.1

**What did you do?**
Hi everyone. I want to make a same vehicle constraint, for example for node A and B, no node order is required. That mean, the result route could be `A -> ... -> B` or `B -> ... -> A` or both dropped. As said [here](https://github.com/google/or-tools/blob/stable/ortools/constraint_solver/routing.h#L869-L884) **AddPickupAndDelivery** should be called for better result in local search phase. But when i use **AddPickupAndDelivery(A, B)**, the solver only allow node A to appear before B.

I'm not sure, may be the document in header file is not updated or there is some bug?

**What did you expect to see**
The solver should allow both `A->B`  or `B -> A`

**What did you see instead?**
Only `A -> B` is valid, `B -> A` make model invalid.

Make sure you include information that can help us debug (full error message, model Proto).

I enclosed an example (Sorry, can't upload a .py file so i rename it to .txt). Node should be on same vehicle is 3 and 4 and **AddPickupAndDelivery(3, 4)** is called. Initial route [1, 3, 4, 2] is valid but initial route [1, 4, 3, 2] isn't.
[main.txt](https://github.com/google/or-tools/files/13246225/main.txt)

**Anything else we should know about your project / environment**

**Comment by ghost** [comment]
Hello, the solver's behavior is intentional, this is not a bug in the documentation.
For a pickup and delivery case, the pickup and the delivery should be on the same vehicle, the pickup appearing before the delivery.
The solver will not work the best if you enforce the constraint you describe with VehicleVar(A) == VehicleVar(B). Instead of a hard constraint, you could add a soft penalty cost if VehicleVar(A) != VehicleVar(B).

**Comment by vuongdanghuy** [comment]
If that’s is intentional, i think the document should be clearer for developer. Because when i read it, it does not say anything about the node order. It just said that method should be use when there is a same vehicle constraint and Pickup - delivery is just an example that could use this method.
